packages/flutter_tools/lib/src/build_system/source.dart

```diff
@@ -0,0 +1,181 @@
+/// [Environment].
+typedef InputFunction = List<FileSystemEntity> Function(
+    Environment environment);
+
+/// Collects sources for a [Target] into a single list of [FileSystemEntities].
+class SourceVisitor {
+  /// Create a new [SourceVisitor] from an [Environment].
+  SourceVisitor(this.environment, [this.inputs = true]);
+
+  static const String _kProjectDirectory = '{PROJECT_DIR}';
+  static const String _kBuildDirectory = '{BUILD_DIR}';
+  static const String _kCacheDirectory = '{CACHE_DIR}';
+  static const String _kPlatform = '{platform}';
+  static const String _kMode = '{mode}';
+  static const String _kFlavor = '{flavor}';
+
+  /// The current environment.
+  final Environment environment;
+
+  /// Whether we are visiting inputs or outputs.
+  ///
+  /// Defaults to `true`.
+  final bool inputs;
+
+  /// The entities are populated after visiting each source.
+  final List<FileSystemEntity> sources = <FileSystemEntity>[];
+
+  /// Visit a [Source] which contains a function.
+  ///
+  /// The function is expected to produce a list of [FileSystemEntities]s.
+  void visitFunction(InputFunction function) {
+    sources.addAll(function(environment));
+  }
+
+  /// Visit a [Source] which contains a file uri.
+  ///
+  /// The uri may that may include constants defined in an [Environment].
+  void visitPattern(String pattern) {
+    // perform substitution of the environmental values and then
+    // of the local values.
+    final List<String> segments = <String>[];
+    final List<String> rawParts = pattern.split('/');
+    final bool hasWildcard = rawParts.last.contains('*');
+    String wildcardFile;
+    if (hasWildcard) {
+      wildcardFile = rawParts.removeLast();
+    }
+    final bool isDirectory = pattern.endsWith('/');
+    // If the pattern does not start with an env variable, then we have nothing
+    // to resolve it to, error out.
+    switch (rawParts.first) {
+      case _kProjectDirectory:
+        segments.addAll(fs.path.split(environment.projectDir.absolute.path));
+        break;
+      case _kBuildDirectory:
+        segments.addAll(fs.path.split(environment.buildDir.absolute.path));
+        break;
+      case _kCacheDirectory:
+        segments.addAll(fs.path.split(environment.cacheDir.absolute.path));
+        break;
+      default:
+        throw InvalidPatternException(pattern);
+    }
+    for (String rawPart in rawParts.skip(1)) {
+      final String value = rawPart
+          .replaceAll(_kMode, getNameForBuildMode(environment.buildMode))
+          .replaceAll(_kPlatform, getNameForTargetPlatform(environment.targetPlatform))
+          .replaceAll(_kFlavor, environment.flavor);
+      segments.add(value);
+    }
+    final String filePath = fs.path.joinAll(segments);
+    if (isDirectory) {
+      sources.add(fs.directory(fs.path.normalize(filePath)));
+    } else if (hasWildcard) {
+      // Perform a simple match by splitting the wildcard containing file on
```

Comment:
Not currently - when you ask glob for a list of files it traverses the real file system instead of our injected filesystem. Separately I could work on the ability to inject the filesystem for glob, which would let us use it in the future.